UCS/SYS New event_set API for using in UCT layer.

hiroyuki-sato opened the pull request:
## What

This PR is an extension event_set API. See also #3682 

## Why ?

It requires to use this API in uct/tcp.

swx-jenkins3 commented:
Can one of the admins verify this patch?

hiroyuki-sato commented on src/ucs/sys/event_set.c:
```diff
@@ -134,36 +135,57 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+                 * but it is ok to not od this */
+                continue;
+            }
+//            ucs_error("epoll_wait() failed: %m");
```
How to care errors except `EINTR`. What do you think to return negative value(-1) in that case?

dmitrygladkov replied:
it should return `0` as a function return value, and we can do something like:
```
unsigned ucs_event_set_wait(ucs_sys_event_set_t *event_set, int max_events,
                            int timeout_ms,
                            ucs_event_set_handler_t event_set_handler,
                            void *arg, ucs_status_t *status_p)
```
...
```
ucs_error(...);
if (status_p != NULL) {
    *status_p = UCS_ERR_IO_ERROR;
}
```

if a user needs to handle errors, it passes storage for `status`, otherwise - `NULL`

hiroyuki-sato replied:
What do you think to return `UCS_OK` on success instead of NULL.
(Or reset NULL on the beginning function.)

I use `status` value on multiple places.
I'm worrying to use the previus result.

```C
    status = ucs_event_set_add(event_set, pipefd[0], UCS_EVENT_SET_EVREAD,
                               NULL);
    EXPECT_EQ(UCS_OK,status);

    nread = ucs_event_set_wait(event_set, MAX_EVENT_SET_SIZE, 20,
                               event_set_func3, NULL, &status);
```

dmitrygladkov replied:
I'm so sorry for being unclear
I meant that we should set `UCS_OK` on success and `UCS_ERR_IO_ERROR` in case of any error occurred if user specified `status_p`, otherwise, we should do nothing

hiroyuki-sato commented on src/ucs/sys/event_set.c:
```diff
@@ -134,36 +135,57 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+        }
+
+        read_events += nready;
+    } while ((read_events < max_events) && (nready == max_wait_events));
```
It based on your codes. https://github.com/openucx/ucx/issues/3652#issuecomment-500110057
I used `max_events` instead of `iface->config.max_poll` 

I worried about misunderstanding your recommendation. Please let me know if I need to fix it.

hiroyuki-sato replied:
fixed

brminich commented:
ok to test

hiroyuki-sato commented on src/ucs/sys/event_set.c:
```diff
@@ -134,36 +135,57 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+    do {
+        max_wait_events = ucs_min(max_events - read_events,
+                                   UCS_EVENT_EPOLL_MAX_EVENTS);
+	nready = epoll_wait(event_set->epfd, ep_events,
```
Fixed

dmitrygladkov replied:
hmm, it seems that it wasn't fixed

hiroyuki-sato replied:
I fixed this part. (alignment)

```C
        max_wait_events = ucs_min(max_events - read_events,
                                  UCS_EVENT_EPOLL_MAX_EVENTS);
```

hiroyuki-sato commented on src/ucs/sys/event_set.c:
```diff
@@ -134,36 +135,57 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+	if (nready < 0) {
+            if (errno == EINTR) {
+                /* here we should take care about timeout_ms (i.e. decrease it),
+                 * but it is ok to not od this */
```
Add `TODO:` and `od` -> `do`.

hiroyuki-sato commented on src/ucs/sys/event_set.c:
```diff
@@ -134,36 +135,57 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+        ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
+                       event_set->epfd, timeout_ms, nready);
+
+        for (i=0; i < nready; i++) {
```
fixed

hiroyuki-sato commented on src/ucs/sys/event_set.c:
```diff
@@ -134,36 +135,57 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+                       event_set->epfd, timeout_ms, nready);
+
+        for (i=0; i < nready; i++) {
+            int events = ucs_event_set_map_to_events(ep_events[i].events);
```
fixed

hiroyuki-sato commented on src/ucs/sys/event_set.c:
```diff
@@ -134,36 +135,57 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
     ucs_free(event_set);
 }
+
+ucs_status_t ucs_event_set_fd_get(ucs_sys_event_set_t *event_set,int *fd_p)
```
fixed

dmitrygladkov commented on test/gtest/ucs/test_event_set.cc:
```diff
@@ -128,11 +135,14 @@ UCS_TEST_F(test_event_set, ucs_event_set_write_thread) {
+    nread = ucs_event_set_wait(event_set, MAX_EVENT_SET_SIZE, 50,
+                               event_set_func2, NULL, &status);
+    EXPECT_EQ(1, nread);
+//    EXPECT_EQ(NULL, &status);
```
pls, remove

hiroyuki-sato replied:
I checked UCS_STATUS instead of NULL check

dmitrygladkov commented on test/gtest/ucs/test_event_set.cc:
```diff
@@ -163,11 +174,14 @@ UCS_TEST_F(test_event_set, ucs_event_set_tmo_thread) {
+    nread = ucs_event_set_wait(event_set, MAX_EVENT_SET_SIZE, 20,
+                               event_set_func3, NULL, &status);
+    EXPECT_EQ(0, nread);
+//    EXPECT_EQ(NULL, status);
```
pls, remove

hiroyuki-sato replied:
I checked UCS_STATUS instead of NULL check

dmitrygladkov commented:
@hiroyuki-sato 

now everything but [this wrong indentation](https://github.com/openucx/ucx/pull/3683/files#r291836404) is OK for me
pls, fix the outstanding comment and mark PR as ready for review

Thank you!

hiroyuki-sato commented:
Hello, @dmitrygladkov 

Thank you for the review.

I fixed indentations. (tab -> space). And rebased.
`ucs_event_set_wait` set `UCS_OK` on success.

Please take a look when you get a chance.

dmitrygladkov commented:
Hello @hiroyuki-sato 

Thank you for fixing comments!
LGTM 👍 

one minor comment: could you pls fix the title of the PR? `UCS/SYS: Extend event_set API`

@brminich @yosefe could you review pls?

hiroyuki-sato commented:
Hello, @dmitrygladkov 
Thank you for the comment. I changed the title

dmitrygladkov commented:
> Hello, @dmitrygladkov
> Thank you for the comment. I changed the title

@hiroyuki-sato Thank you!

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10571/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10572/ for details (Mellanox internal link).

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10573/ for details (Mellanox internal link).

brminich commented on test/gtest/ucs/test_event_set.cc:
```diff
@@ -93,10 +96,13 @@ UCS_TEST_F(test_event_set, ucs_event_set_read_thread) {
 
-    status = ucs_event_set_add(event_set, pipefd[0], UCS_EVENT_SET_EVREAD);
+    status = ucs_event_set_add(event_set, pipefd[0], UCS_EVENT_SET_EVREAD,
+                               (void *)&pipefd[0]);
```
minor: can pass by value

hiroyuki-sato replied:
Does this change is your suggestion?

```diff
     status = ucs_event_set_add(event_set, pipefd[0], UCS_EVENT_SET_EVREAD,
-                               (void *)&pipefd[0]);
+                               (void *)(uintptr_t)pipefd[0]);
```

brminich commented on src/ucs/sys/event_set.c:
```diff
@@ -134,36 +135,58 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+        }
+
+        read_events += nready;
+    } while ((read_events < max_events) && (nready == max_wait_events));
```
why need this cycle? Can we go with just one successful iteration?

dmitrygladkov replied:
Hello @hiroyuki-sato 

We discussed this with @brminich and decided that's would be better to manage responsibilities between layers in a different way.

This loop will affect UCS/ASYNC/THREAD, so the loop should be done in TCP
here we should take care about timeout that is needed for UCS/ASYNC

So, we propose the following change:
```
ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set,
                                unsigned max_events, int timeout_ms,
                                ucs_event_set_handler_t event_set_handler,
                                void *arg, unsigned *read_events)
{
    struct epoll_event ep_events[UCS_EVENT_EPOLL_MAX_EVENTS];
    unsigned i, max_wait_events;
    int nready, events;

    ucs_assert(event_set_handler != NULL);

    max_wait_events = ucs_min(max_events, UCS_EVENT_EPOLL_MAX_EVENTS);
    nready = epoll_wait(event_set->epfd, ep_events, max_wait_events, timeout_ms);
    if (nready < 0) {
        *read_events = 0;

        if (errno == EINTR) {
            return UCS_INPROGRESS;
        }

        ucs_fatal("epoll_wait() failed: %m");
        return UCS_ERR_IO_ERROR;
    }
    
    ucs_trace_data("epoll_wait(epfd=%d, timeout=%d) returned %d",
                   event_set->epfd, timeout_ms, nready);

    for (i = 0; i < nready; i++) {
        events = ucs_event_set_map_to_events(ep_events[i].events);
        event_set_handler(ep_events[i].data.ptr, events, arg);
    }

    *read_events = nready;
    if  ((nready < max_events) && (nready == UCS_EVENT_EPOLL_MAX_EVENTS)) {
        return UCS_INPROGRESS;
    }

    return UCS_OK;
}
```
```
/**
 * Wait for an I/O events
 *
 * @param [in]  event_set          Event set created by ucs_event_set_create.
 * @param [in]  max_events         Maximum wait events.
 * @param [in]  timeout_ms         Timeout period in ms.
 * @param [in]  event_set_handler  Callback functions.
 * @param [in]  arg                User data variables.
 * @param [out] read_events        number of read events.
 *
 * @return return UCS_OK on success, UCS_INPROGRESS - call was interrupted by a
 *         signal handler or there are probably more events to read,
 *         UCS_ERR_IO_ERROR - an error occurred.
 */
ucs_status_t ucs_event_set_wait(ucs_sys_event_set_t *event_set,
                                unsigned max_events, int timeout_ms,
                                ucs_event_set_handler_t event_set_handler,
                                void *arg, unsigned *read_events);
```

hiroyuki-sato replied:
I modified ucs_event_set_wait with your proposed way.

dmitrygladkov commented:
@hiroyuki-sato Thank you for the updates!

dmitrygladkov commented on test/gtest/ucs/test_event_set.cc:
```diff
@@ -93,10 +96,13 @@ UCS_TEST_F(test_event_set, ucs_event_set_read_thread) {
-    status = ucs_event_set_wait(event_set, 50, event_set_func1, arg);
+    status = ucs_event_set_wait(event_set, MAX_EVENT_SET_SIZE, 50,
+                                event_set_func1, arg, &nread);
+    EXPECT_EQ(1, nread);
```
`EXPECT_EQ(1u, nread);`

dmitrygladkov commented on test/gtest/ucs/test_event_set.cc:
```diff
@@ -128,10 +135,13 @@ UCS_TEST_F(test_event_set, ucs_event_set_write_thread) {
-    status = ucs_event_set_wait(event_set, 50, event_set_func2, NULL);
+    status = ucs_event_set_wait(event_set, MAX_EVENT_SET_SIZE, 50,
+                                event_set_func2, NULL, &nread);
+    EXPECT_EQ(1, nread);
```
`EXPECT_EQ(1u, nread);`

dmitrygladkov commented on test/gtest/ucs/test_event_set.cc:
```diff
@@ -163,10 +174,13 @@ UCS_TEST_F(test_event_set, ucs_event_set_tmo_thread) {
-    status = ucs_event_set_wait(event_set, 20, event_set_func3, NULL);
+    status = ucs_event_set_wait(event_set, MAX_EVENT_SET_SIZE, 20,
+                                event_set_func3, NULL, &nread);
+    EXPECT_EQ(0, nread);
```
`EXPECT_EQ(0u, nread);`

dmitrygladkov commented:
I've added comments to fix the following error:

```
14:48:53 In file included from /scrap/jenkins/workspace/hpc-ucx-pr-4/label/hpc-test-node/worker/1/contrib/../test/gtest/common/test_helpers.h:11,
14:48:53                  from /scrap/jenkins/workspace/hpc-ucx-pr-4/label/hpc-test-node/worker/1/contrib/../test/gtest/common/test.h:10,
14:48:53                  from /scrap/jenkins/workspace/hpc-ucx-pr-4/label/hpc-test-node/worker/1/contrib/../test/gtest/ucs/test_event_set.cc:7:
14:48:53 /scrap/jenkins/workspace/hpc-ucx-pr-4/label/hpc-test-node/worker/1/contrib/../test/gtest/common/gtest.h: In instantiation of ‘testing::AssertionResult testing::internal::CmpHelperEQ(const char*, const char*, const T1&, const T2&) [with T1 = int; T2 = unsigned int]’:
14:48:53 /scrap/jenkins/workspace/hpc-ucx-pr-4/label/hpc-test-node/worker/1/contrib/../test/gtest/common/gtest.h:18898:23:   required from ‘static testing::AssertionResult testing::internal::EqHelper<lhs_is_null_literal>::Compare(const char*, const char*, const T1&, const T2&) [with T1 = int; T2 = unsigned int; bool lhs_is_null_literal = false]’
14:48:53 /scrap/jenkins/workspace/hpc-ucx-pr-4/label/hpc-test-node/worker/1/contrib/../test/gtest/ucs/test_event_set.cc:105:5:   required from here
14:48:53 /scrap/jenkins/workspace/hpc-ucx-pr-4/label/hpc-test-node/worker/1/contrib/../test/gtest/common/gtest.h:18862:16: error: comparison of integer expressions of different signedness: ‘const int’ and ‘const unsigned int’ [-Werror=sign-compare]
14:48:53    if (expected == actual) {
14:48:53        ~~~~~~~~~^~~~~~~~~
14:48:54   CXX      ucp/gtest-test_ucp_stream.o
14:48:54   CXX      ucp/gtest-test_ucp_peer_failure.o
14:48:54   CXX      ucp/gtest-test_ucp_atomic.o
14:48:54 cc1plus: all warnings being treated as errors
```

hiroyuki-sato commented:
Hello, @dmitrygladkov 
Thank you for the comment. I fixed these comments.
Do you know whether is it possible to check on the local machine or not?

I executed `GTEST_FILTER='*test_event_set*' make -C test/gtest test` before push it.

dmitrygladkov commented:
> Hello, @dmitrygladkov
> Thank you for the comment. I fixed these comments.
> Do you know whether is it possible to check on the local machine or not?
> 
> I executed `GTEST_FILTER='*test_event_set*' make -C test/gtest test` before push it.

Hello @hiroyuki-sato 

It is possible to reproduce this compiler error w/ gcc 8.2.0

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10585/ for details (Mellanox internal link).

swx-jenkins commented:
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/7617/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10591/ for details (Mellanox internal link).

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/7627/ for details.

brminich commented:
build time-out

bot:mlx:retest

brminich commented on src/ucs/sys/event_set.c:
```diff
@@ -134,31 +135,43 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
-    if (event_set_handler == NULL) {
-        return UCS_ERR_INVALID_PARAM;
-    }
+    ucs_assert(event_set_handler != NULL);
```
plz, check read_events as well

hiroyuki-sato replied:
Added ucs_assert();

brminich commented on src/ucs/sys/event_set.c:
```diff
@@ -134,31 +135,43 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+        if (errno == EINTR) {
+            return UCS_INPROGRESS;
+        }
+        ucs_fatal("epoll_wait() failed: %m");
```
error would be enough, fatal cause abort

hiroyuki-sato replied:
Changed to `ucs_error()`

brminich commented on src/ucs/sys/event_set.h:
```diff
@@ -9,6 +9,8 @@
 
 #include <ucs/type/status.h>
 
+#define UCS_EVENT_SET_MAX_POLL_SIZE 16
```
where are you going to use it?

hiroyuki-sato replied:
Removed. Old code used.

hiroyuki-sato commented:
Hello, @dmitrygladkov 

> It is possible to reproduce this compiler error w/ gcc 8.2.0

OMG 

```
gcc --version
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-36)
```

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10598/ for details (Mellanox internal link).

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/7632/ for details.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10600/ for details (Mellanox internal link).

hiroyuki-sato commented:
Hello, @brminich and @dmitrygladkov 

Thank you for the review.

I've set wrong commit message prefix. It missing colon.
Now: `UCS/SYS` Correct: `UCS/SYS:`

Should I change commit message?, If so Should I rebase into one commit?

If so, do I need to keep this commit message?

```
Before this change, the loop in ucs_event_set_wait() will affect
UCS/ASYNC/THREAD, so the loop should be done in TCP here we
should take care about timeout that is needed for UCS/ASYNC
```

Best regards.

brminich commented:
Hello @hiroyuki-sato,
Better squash the commits, as it is self-contained API change (and you could add the colon).
The mentioned commit message is not needed, since it describes the old/obsolete code.

hiroyuki-sato commented:
Hello, @brminich. Thank you for your reply. I squashed the commits!

brminich commented:
Thanks, looks good to me

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/7633/ for details.

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10601/ for details (Mellanox internal link).

dmitrygladkov commented:
bot:mlx:retest

mellanox-github commented:
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10605/ for details (Mellanox internal link).

hiroyuki-sato commented:
Does this test failed on vulcan-jenkins,0?, It this correct?
Could you tell me how to confirm the error detail, if this error caused by this PR?

![スクリーンショット 2019-06-12 9 15 05](https://user-images.githubusercontent.com/767650/59315045-f4e2f680-8cf2-11e9-9cd7-ff7f1d88b618.png)

dmitrygladkov commented:
> Does this test failed on vulcan-jenkins,0?, It this correct?

Yes, there is a failure on vulcan-jenkins,0.

```
14:14:57 [ RUN      ] thread_spinlock/test_async.ctx_event/0
14:14:57 /scrap/jenkins/workspace/hpc-ucx-pr/label/vulcan-jenkins/worker/0/contrib/../test/gtest/ucs/test_async.cc:451: Failure
14:14:57 Expected: (le.count()) >= (1), actual: 0 vs 1
```

> Could you tell me how to confirm the error detail, if this error caused by this PR?

let's try to retest, because it may be unrelated to you changes (the previous testing was successful for this test).

dmitrygladkov commented:
bot:mlx:retest

hiroyuki-sato commented:
Hello, @dmitrygladkov 
Thank you for the comment.

Can I see the following log in my environment? or Mellanox internal only?
If I can, could you tell me what page should I check?
(I'm newbie a Jenkins.)

And can I execute rerun command using `bot:mlx...` ?
(Maybe, it is Mellanox staff only. Just confirmation.)

Best reagards.

```
14:14:57 [ RUN      ] thread_spinlock/test_async.ctx_event/0
14:14:57 /scrap/jenkins/workspace/hpc-ucx-pr/label/vulcan-jenkins/worker/0/contrib/../test/gtest/ucs/test_async.cc:451: Failure
14:14:57 Expected: (le.count()) >= (1), actual: 0 vs 1
```

dmitrygladkov commented:
Hello @hiroyuki-sato 

> Can I see the following log in my environment? or Mellanox internal only?
> If I can, could you tell me what page should I check?
> (I'm newbie a Jenkins.)

This is Mellanox internal only

> And can I execute rerun command using bot:mlx... ?
> (Maybe, it is Mellanox staff only. Just confirmation.)

Looks like you can do it

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10614/ for details (Mellanox internal link).

hiroyuki-sato commented:
Hello, @dmitrygladkov Thank you for your reply. Test Passed!

>> And can I execute rerun command using bot:mlx... ?
>> (Maybe, it is Mellanox staff only. Just confirmation.)

> Looks like you can do it

I'll try next time.

yosefe commented on src/ucs/sys/event_set.c:
```diff
@@ -134,31 +135,44 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
-    for (i=0; i < nready; i++) {
-        int events = ucs_event_set_map_to_events(ep_events[i].events);
-        event_set_handler(ep_events[i].data.fd, events, arg);
+    for (i = 0; i < nready; i++) {
```
there is signed/unsigned compare here (some compilers may not like it)

hiroyuki-sato replied:
Changed `i` from `unsigned` to `signed`.

yosefe commented on src/ucs/sys/event_set.c:
```diff
@@ -134,31 +135,44 @@ ucs_status_t ucs_event_set_del(ucs_sys_event_set_t *event_set, int event_fd)
+
+        if (errno == EINTR) {
+            return UCS_INPROGRESS;
+        }
```
space line after

hiroyuki-sato replied:
Changed 

```C
        if (errno == EINTR) {
            return UCS_INPROGRESS;
        }

        ucs_error("epoll_wait() failed: %m");
        return UCS_ERR_IO_ERROR;
```

yosefe commented on src/ucs/sys/event_set.c:
```diff
@@ -167,3 +181,13 @@ void ucs_event_set_cleanup(ucs_sys_event_set_t *event_set)
+         return UCS_ERR_INVALID_PARAM;
+    }
+    *fd_p = event_set->epfd;
+
```
this space line should be moved before line 190

hiroyuki-sato replied:
Changed below.

```C
     if (event_set == NULL) {
          return UCS_ERR_INVALID_PARAM;
     }

    *fd_p = event_set->epfd;
     return UCS_OK;
```

yosefe commented on src/ucs/sys/event_set.h:
```diff
@@ -20,7 +20,8 @@ typedef struct ucs_sys_event_set ucs_sys_event_set_t;
  *
  */
-typedef void (*ucs_event_set_handler_t)(int fd, int event, void *arg);
+typedef void (*ucs_event_set_handler_t)(void *callback_data, int event,
```
pls add documentation of callback arguments

hiroyuki-sato replied:
Added doxygen comments.

```C
/**
 * ucs_event_set_handler call this handler for notifying event
 *
 * @param [in] callback_data  User data which set in ucs_event_set_add().
 * @param [in] event          Detection event. Sets of ucs_event_set_type_t.
 * @param [in] arg            User data which set in ucs_event_set_wait().
 *
 */
```

swx-jenkins commented:
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/7649/ for details.

mellanox-github commented:
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10622/ for details (Mellanox internal link).

yosefe commented:
bot:pipe:retest

hiroyuki-sato commented:
Thank you.
